azurelinuxagent/ga/cgroupapi.py

```diff
@@ -342,11 +440,85 @@ def start_extension_command(self, extension_name, command, cmd_name, timeout, sh
 
-    def cleanup_legacy_cgroups(self):
+
+class SystemdCgroupApiv2(_SystemdCgroupApi):
+    """
+    Cgroup v2 interface via systemd
+    """
+    def __init__(self):
+        super(SystemdCgroupApiv2, self).__init__()
+        self._root_cgroup_path = self._get_root_cgroup_path()
+        self._controllers_enabled_at_root = self._get_controllers_enabled_at_root()
+
+    @staticmethod
+    def _get_root_cgroup_path():
         """
-        Previous versions of the daemon (2.2.31-2.2.40) wrote their PID to /sys/fs/cgroup/{cpu,memory}/WALinuxAgent/WALinuxAgent;
-        starting from version 2.2.41 we track the agent service in walinuxagent.service instead of WALinuxAgent/WALinuxAgent. If
-        we find that any of the legacy groups include the PID of the daemon then we need to disable data collection for this
-        instance (under systemd, moving PIDs across the cgroup file system can produce unpredictable results)
+        In v2, there is a unified mount point shared by all controllers. Use findmnt to get the unified mount point.
+
+          The output of findmnt is similar to
+              $ findmnt -t cgroup2 --noheadings
+              /sys/fs/cgroup cgroup2 cgroup2 rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot
+
+        Returns None if the root cgroup cannot be determined from the output above.
+        """
+        #
+        for line in shellutil.run_command(['findmnt', '-t', 'cgroup2', '--noheadings']).splitlines():
+            # Systemd mounts the cgroup filesystem at '/sys/fs/cgroup'. The agent does not support cgroups if the
+            # filesystem is mounted elsewhere, so search specifically for '/sys/fs/cgroup' in the findmnt output.
+            match = re.search(r'(?P<path>\/sys\/fs\/cgroup)\s+cgroup2', line)
+            if match is not None:
+                root_cgroup_path = match.group('path')
+                if root_cgroup_path is not None:
+                    return root_cgroup_path
+        return None
+
+    def _get_controllers_enabled_at_root(self):
+        """
+        Returns a list of the controllers enabled at the root cgroup. The cgroup.subtree_control file at the root shows
+        a space separated list of the controllers which are enabled to control resource distribution from the root
+        cgroup to its children. If a controller is listed here, then that controller is available to enable in children
+        cgroups.
+
+                $ cat /sys/fs/cgroup/cgroup.subtree_control
+                cpuset cpu io memory hugetlb pids rdma misc
         """
-        return CGroupsApi._foreach_legacy_cgroup(lambda *_: None)
+        controllers_enabled_at_root = []
+        if self._root_cgroup_path is not None:
+            enabled_controllers_file = os.path.join(self._root_cgroup_path, 'cgroup.subtree_control')
+            if os.path.exists(enabled_controllers_file):
+                controllers_enabled_at_root = fileutil.read_file(enabled_controllers_file).rstrip().split(" ")
```

Comment:
This only handle single space on split, should we consider multiple whitespaces and just call split without delimiter?

Reply:
systemd mentions it is a space separated list. Anything else would be unexpected. 

I'll remove delimiter to be safe